### PR TITLE
Added methods to calculate and plot: mutations-per-tree…

### PR DESCRIPTION
**Pull request by savitakartik**
… distribution, sites-per-tree distribution, mutations per tree along the sequence and sites per tree along the sequence.

Closes issue #8

### CODE REVIEW

**Comment by savitakartik** [comment]
I haven't implemented tree.num_mutations differently (as mentioned in #8), but the distribution  for the short arm of chr2 (Unified Genealogies ts) takes 35 seconds to plot.

**Comment by jeromekelleher** [comment]
Thanks - can you paste in example plots please?

**Comment by savitakartik** [comment]
> Thanks - can you paste in example plots please?

Sorry! Sure, for mutations per tree:
image
image
Sites per tree:
image
image

**Comment by jeromekelleher** [comment]
Love it!

File: utils.py [hunk 1]
```diff
@@ -291,3 +290,142 @@ def plot_mean_node_arity(self, show_counts=False):
         )
         if show_counts:
             plt.bar_label(bars, fmt="{:,.0f}")
+
+    def calc_mutations_per_tree(self):
+        tree_mutations = np.zeros(self.ts.num_trees)
+        for tree in self.ts.trees():
+            tree_mutations[tree.index] = tree.num_mutations
+        return tree_mutations
```
Comment (benjeffery) [inline utils.py:298]: I added a comment on https://github.com/tskit-dev/tsinfer-qc/issues/8#issuecomment-1583549582 suggesting an efficient method for calculating mutations per tree.

File: utils.py [hunk 2]
```diff
@@ -291,3 +290,142 @@ def plot_mean_node_arity(self, show_counts=False):
...
+            trees_with_many_muts = np.sum(
+                self.calc_mutations_per_tree() > max_num_muts)
+            ax.text(
+                0.01,
+                0.99,
+                f"there are {trees_with_many_muts:,} trees\nwith more than {max_num_muts:,} mutations",
+                transform=ax.transAxes,
+            )
+        counts, edges, bars = plt.hist(
+            self.calc_mutations_per_tree(), bins=bins, edgecolor="black"
```
Comment (benjeffery) [inline utils.py:314]: As it is expensive, there's no need to call `self.calc_mutations_per_tree` twice here and on 306.

File: utils.py [hunk 3]
```diff
@@ -291,3 +290,142 @@ def plot_mean_node_arity(self, show_counts=False):
...
+        grid.ax_marg_y.remove()
+        grid.fig.set_figwidth(20)
+        grid.fig.set_figheight(8)
+        grid.ax_joint.set_xlabel("Position on genome (Mb)")
+        grid.ax_joint.set_ylabel("Number of mutations per tree")
+
+    def calc_sites_per_tree(self):
+        tree_sites = np.zeros(self.ts.num_trees)
+        for tree in self.ts.trees():
+            tree_sites[tree.index] = tree.num_sites
```
Comment (benjeffery) [inline utils.py:366]: The comment on #8 has a quick way to get site counts too.

**Review by benjeffery — COMMENTED** [review]
Looks good, couple of comments.

I think the default bins on the histos are pretty wide, maybe something like `max(100,sqrt(N))` would give nice bins?

**Review by jeromekelleher — COMMENTED** [review]
LGTM - @benjeffery has some good feedback here on storing the results of expensive methods. Other than that we can merge and plug in a more efficient approach later?

File: utils.py (hunk 3 above)
Comment (savitakartik) [reply utils.py]: I've made these changes now. Plotting mutations per tree takes <1 second now (chr 2q UG). Thanks a lot, @benjeffery .

**Pull request merged**